### PR TITLE
feat: add node spec to the persistent store

### PR DESCRIPTION
**Pull request by tiagolobocastro**
Write out the node spec to the persistent store.
On startup the core agent can use this to attempt to load all node information if the
nodes are online.

Changed the openapi spec to return the Node with a spec and state.
Rename the state/status for the nodes properly.

### CODE REVIEW

File: control-plane/agents/core/src/core/wrapper.rs [hunk 1]
```diff
@@ -217,17 +248,19 @@ impl NodeWrapper {
...
             }
         } else {
             tracing::trace!(
                 "Skipping reload of node '{}' since it's '{:?}'",
                 self.id,
-                self.state
+                self.status
             );
+            let mut states = self.states.write();
+            states.clear_all();
```
Comment (paulyoong) [inline control-plane/agents/core/src/core/wrapper.rs:263]: Nit: You could do `self.states.write().clear_all()` like you do on 251.

File: control-plane/agents/core/src/node/registry.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+use crate::core::{registry::Registry, wrapper::NodeWrapper};
+use common::errors::{NodeNotFound, SvcError};
+use common_lib::types::v0::message_bus::{NodeId, NodeState, Register};
+use snafu::OptionExt;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+impl Registry {
+    /// Get all node wrappers
+    pub(crate) async fn get_nodes_wrapper(&self) -> Vec<Arc<Mutex<NodeWrapper>>> {
```
Comment (paulyoong) [inline control-plane/agents/core/src/node/registry.rs:10]: get_nodes_wrapper -> get_node_wrappers

File: control-plane/agents/core/src/node/specs.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            .nodes
+            .get(node_id)
+            .cloned()
+            .context(NodeNotFound {
+                node_id: node_id.to_owned(),
+            })
+    }
+
+    /// Get cloned node spec by its `NodeId`
+    pub(crate) fn get_node(&self, node_id: &NodeId) -> Result<NodeSpec, SvcError> {
```
Comment (paulyoong) [inline control-plane/agents/core/src/node/specs.rs:58]: get_node -> get_node_spec

File: control-plane/agents/core/src/node/specs.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            }
+        };
+        if let Ok(node) = &node {
+            registry.store_obj(node).await?;
+        }
+        node
+    }
+
+    /// Get node spec by its `NodeId`
+    pub(crate) fn get_locked_node(
```
Comment (paulyoong) [inline control-plane/agents/core/src/node/specs.rs:44]: get_locked_node -> get_locked_node_spec

**Review by paulyoong — APPROVED** [review]

File: common/src/types/v0/message_bus/node.rs [hunk 5]
```diff
@@ -26,13 +27,81 @@ pub struct Deregister {
...
+    /// Runtime state of the node.
+    state: Option<NodeState>,
+}
+impl Node {
+    /// Get new `Self` from the given parameters
+    pub fn new(id: NodeId, spec: Option<NodeSpec>, state: Option<NodeState>) -> Self {
+        Self { id, spec, state }
+    }
+    /// Get new `Self` from the given parameters
+    pub fn new_spec_state(id: NodeId, endpoint: String, status: NodeStatus) -> Self {
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply common/src/types/v0/message_bus/node.rs:66]: Yep, moved to a test function.

File: common/src/types/v0/store/registry.rs [hunk 6]
```diff
@@ -0,0 +1,77 @@
+use crate::types::v0::store::definitions::{ObjectKey, StorableObject, StorableObjectType};
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+
+/// Registry configuration loaded from/stored into the persistent store
+#[derive(Serialize, Deserialize, Debug)]
+pub struct CoreRegistryConfig {
+    /// Key of this configuration
+    id: CoreRegistryConfigKey,
+    /// Node identification.
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply common/src/types/v0/store/registry.rs:10]: Doh

File: control-plane/agents/common/src/errors.rs [hunk 7]
```diff
@@ -170,11 +170,17 @@ pub enum SvcError {
     ReplicaRemovalNoCandidates { id: String },
     #[snafu(display("No online replicas are available for Volume '{}'", id))]
     NoOnlineReplicas { id: String },
+    /// Failed to find an entry with the given key.
+    #[snafu(display("Entry with key {} not found.", key))]
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/agents/common/src/errors.rs:174]: Ah yes indeed

File: control-plane/agents/core/src/core/wrapper.rs [hunk 8]
```diff
@@ -23,7 +23,7 @@ use std::cmp::Ordering;
 #[derive(Debug, Clone)]
 pub(crate) struct NodeWrapper {
     /// inner Node value
-    node: Node,
+    node: NodeState,
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/agents/core/src/core/wrapper.rs:26]: yeah, done

File: control-plane/agents/core/src/core/wrapper.rs (hunk 1 above)
Comment (tiagolobocastro) [reply control-plane/agents/core/src/core/wrapper.rs:263]: :+1:

**Comment by tiagolobocastro** [comment]
bors merge

**Pull request merged**